### PR TITLE
[Merged by Bors] - chore(CategoryTheory/Monoidal): oplax functors and comonoids

### PR DESCRIPTION
**Pull request by kim-em**
Oplax monoidal functors take comonoids to comonoids. Also, generalize constructions of limits in `Mon_` so limits of a particular shape in `C` are enough to give limits of that shape in `Mon_ C`.

- [x] depends on: #13310
- [x] depends on: #13316

### CODE REVIEW

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean [hunk 1]
```diff
@@ -41,106 +42,68 @@ namespace MonFunctorCategoryEquivalence
...
+      one_hom := congr_app f.one_hom X
+      mul_hom := congr_app f.mul_hom X } }
 
--- Porting note: the `obj` field of `inverse : (C ⥤ Mon_ D) ⥤ Mon_ (C ⥤ D)` defined below
--- had to be defined separately as `Inverse.obj` in order to speed up the compilation
 /-- A functor to the category of monoid objects can be translated as a monoid object
 in the functor category. -/
 @[simps]
-def Inverse.obj (F : C ⥤ Mon_ D) : Mon_ (C ⥤ D) where
+private def inverse_obj' (F : C ⥤ Mon_ D) : Mon_ (C ⥤ D) where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean:77]: I do not understand why these definitions are `private`.

File: Mathlib/CategoryTheory/Monoidal/Internal/Limits.lean [hunk 2]
```diff
@@ -86,21 +86,17 @@ def limitConeIsLimit (F : J ⥤ Mon_ C) : IsLimit (limitCone F) where
 set_option linter.uppercaseLean3 false in
 #align Mon_.limit_cone_is_limit Mon_.limitConeIsLimit
 
-instance hasLimits : HasLimits (Mon_ C) where
-  has_limits_of_shape _ _ :=
-    { has_limit := fun F =>
-        HasLimit.mk
-          { cone := limitCone F
-            isLimit := limitConeIsLimit F } }
+instance hasLimits [HasLimitsOfShape J C] : HasLimitsOfShape J (Mon_ C) where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Monoidal/Internal/Limits.lean:89]:
```suggestion
instance hasLimitsOfShape [HasLimitsOfShape J C] : HasLimitsOfShape J (Mon_ C) where
```

File: Mathlib/CategoryTheory/Monoidal/Internal/Limits.lean [hunk 3]
```diff
@@ -86,21 +86,17 @@ def limitConeIsLimit (F : J ⥤ Mon_ C) : IsLimit (limitCone F) where
...
 set_option linter.uppercaseLean3 false in
 #align Mon_.has_limits Mon_.hasLimits
 
-instance forgetPreservesLimits : PreservesLimits (Mon_.forget C) where
-  preservesLimitsOfShape :=
-    { preservesLimit := fun {F} =>
-        preservesLimitOfPreservesLimitCone (limitConeIsLimit F)
-          (IsLimit.ofIsoLimit (limit.isLimit (F ⋙ Mon_.forget C))
-            (forgetMapConeLimitConeIso F).symm) }
+instance forgetPreservesLimits : PreservesLimitsOfShape J (Mon_.forget C) where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Monoidal/Internal/Limits.lean:96]:
```suggestion
instance forgetPreservesLimitsOfShape : PreservesLimitsOfShape J (Mon_.forget C) where
```

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean]: I had been thinking these would never need to be accessed once the equivalence was built, but I agree this is not helpful, and the namespace is already sufficient to indicate these are "implementation details".

Comment (semorrison) [reply Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean]: (Removed now.)

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean [hunk 4]
```diff
@@ -41,22 +42,20 @@ namespace MonFunctorCategoryEquivalence
 
 variable {C D}
 
--- Porting note: the `obj` field of `functor : Mon_ (C ⥤ D) ⥤ C ⥤ Mon_ D` defined below
--- had to be defined separately as `Functor.obj` in order to speed up the compilation
 /-- A monoid object in a functor category induces a functor to the category of monoid objects. -/
 @[simps]
-def Functor.obj (A : Mon_ (C ⥤ D)) : C ⥤ Mon_ D where
+def functor_obj' (A : Mon_ (C ⥤ D)) : C ⥤ Mon_ D where
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean:47]:
```suggestion
def functorObj' (A : Mon_ (C ⥤ D)) : C ⥤ Mon_ D where
```

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean [hunk 5]
```diff
@@ -65,82 +64,46 @@ to a functor into the category of monoid objects.
...
+      one_hom := congr_app f.one_hom X
+      mul_hom := congr_app f.mul_hom X } }
 
--- Porting note: the `obj` field of `inverse : (C ⥤ Mon_ D) ⥤ Mon_ (C ⥤ D)` defined below
--- had to be defined separately as `Inverse.obj` in order to speed up the compilation
 /-- A functor to the category of monoid objects can be translated as a monoid object
 in the functor category. -/
 @[simps]
-def Inverse.obj (F : C ⥤ Mon_ D) : Mon_ (C ⥤ D) where
+def inverse_obj' (F : C ⥤ Mon_ D) : Mon_ (C ⥤ D) where
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean:77]:
```suggestion
def inverseObj' (F : C ⥤ Mon_ D) : Mon_ (C ⥤ D) where
```

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean [hunk 6]
```diff
@@ -159,6 +122,92 @@ def monFunctorCategoryEquivalence : Mon_ (C ⥤ D) ≌ C ⥤ Mon_ D where
...
 
+namespace ComonFunctorCategoryEquivalence
+
+variable {C D}
+
+/--
+A comonoid object in a functor category induces a functor to the category of comonoid objects.
+-/
+@[simps]
+private def functor_obj' (A : Comon_ (C ⥤ D)) : C ⥤ Comon_ D where
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean:133]: Should this be private or not? I assume we either both or neither of the `Mon/Comon` versions to be private.

Similarly for all the other `private` decls below

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean:133]: Also, `functorObj'` is the proper naming, right?

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean (hunk 6 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Monoidal/Internal/FunctorCategory.lean]: Thanks for the suggestions, I think they are all correct. I've removed `private`.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**